### DISCUSSION TITLE
Save game format of STNN chunk - is it a bit more dynamic than the table header suggests?

### DISCUSSION THREAD

**Discussion by michalc** [OpenTTD/OpenTTD#11361] [Q&A] [answered]
Hello 👋 

I'm writing a parser for the OpenTTD save game format using 
 https://github.com/OpenTTD/OpenTTD/blob/master/docs/savegame_format.md as a guide. It says:

> As each type has a well defined length, you can read the records even without knowing anything about the chunk-tag yourself.

But I'm wondering if the STNN has a bit of extra behaviour here that does require some knowledge beyond the table header. Looking at https://github.com/OpenTTD/OpenTTD/blob/84201a85202a4ddb972f953c97e2d4f90be1c1e1/src/saveload/station_sl.cpp#L749C4-L749C57

```C++
bool waypoint = (SlReadByte() & FACIL_WAYPOINT) != 0;

BaseStation *bst = waypoint ? (BaseStation *)new (index) Waypoint() : new (index) Station();
SlObject(bst, slt);
```
It looks like what comes after the first "facilities" byte is _either_ a Waypoint or normal(?) Station, and not both? So you can't expect a gamma for both to list how many there are.

(I'm no C++ expert, apologies if the answer is obvious)

**Comment by TrueBrain** [comment, marked as answer]
Stations in OpenTTD are a bit weird. But nothing really for you to worry about.

In the header you will see there is a `normal` field and a `waypoint` field, which are both structs. Either of those two are filled, never both.

The line you point to is a trick we use when loading the savegame to split stations from waypoints. But this is purely a C++ thingy, and has no reflections on the savegame itself.

Except for some chunks that have no table header, you will see you can actually fully load the savegame based on the information in the header; no need to worry about how we load it in C++.

There is a reference implementation here:
https://github.com/TrueBrain/OpenTTD-savegame-reader
You can check out the result here:
https://truebrain.github.io/OpenTTD-savegame-reader/

You can use this reference implementation to check if information can be decoded: if it shows up there, the headers contain enough information to decode it :)

If you have any other questions or get stuck, feel free to ask!

**Reply by michalc** [reply]
Ah understood, and....

> There is a reference implementation here:
https://github.com/TrueBrain/OpenTTD-savegame-reader

This looks great! So my aim isn't actually the parser, but to extract data from the games... I am kinda tempted to abandon mine and use yours. Thanks!
